Add metrics builder (#438)
axiomhq/docs: query-data/metrics/overview.mdx

---
title: Metrics
description: This section explains how to work with OpenTelemetry (OTel) metrics.
sidebarTitle: Overview
---

import Prerequisites from "/snippets/standard-prerequisites.mdx"

Axiom supports ingesting, querying, and visualizing OpenTelemetry (OTel) metrics alongside logs and traces. This section explains how to work with OTel metrics in Axiom.

<Info>
Support for OTel metrics is currently in private preview. For more information, see [Feature states](/platform-overview/roadmap#feature-states).
</Info>

> <Prerequisites />

Reviewer: I realise this isn't being introduced by this PR, but it might be confusing that we're advising users that their token should have "permissions to update the dataset" in order to ingest data. We do have a 'Dataset' row in the permissions table with an 'Update' checkbox, but this covers update actions to the resource, for example changing the description. To ingest, you want to ensure the explicit ingest permission is granted.

Author: Thanks! Fixing in https://github.com/axiomhq/docs/pull/455

---
title: Metrics
description: This section explains how to work with OpenTelemetry (OTel) metrics.
sidebarTitle: Overview
---

import Prerequisites from "/snippets/standard-prerequisites.mdx"

Axiom’s dedicated MetricsDB provides a purpose-built metrics database that handles high-cardinality time-series data without the cost penalties and performance degradation common in traditional metrics systems. This section explains how to work with OpenTelemetry (OTel) metrics in Axiom.

<Info>
Support for OTel metrics is currently in private preview. For more information, see [Feature states](/platform-overview/roadmap#feature-states).

Axiom is confident in the quality of MetricsDB and the reliability of the data it stores. However, preview features are experimental, and Axiom doesn’t guarantee data integrity or accuracy for preview features.
</Info>

## What makes MetricsDB different

MetricsDB is engineered from the ground up to embrace dimensional complexity:

> - **High-cardinality as a design principle**: Store metrics with high-cardinality dimensional tags. Where other metrics databases penalize you with higher costs or degraded performance, MetricsDB treats high cardinality as a core capability.

Reviewer: ```suggestion
- **High-cardinality as a design principle**: Store metrics with high-cardinality tags. Where other metrics databases penalize you with higher costs or degraded performance, MetricsDB treats high cardinality as a core capability.
```

</Frame>

## Design choices and constraints

MetricsDB makes intentional architectural trade-offs to optimize for the most common metrics use cases while maintaining exceptional performance at scale.

### Query scope

You can query one dataset per query.

### Supported data types

MetricsDB focuses on the core OpenTelemetry metric types that cover the vast majority of observability scenarios.

Axiom supports the following OpenTelemetry metric types:
- **Counter**: Monotonically increasing values. For example, request count.
- **UpDownCounter**: Values that can increase or decrease. For example, active connections.
- **Gauge**: Point-in-time measurements. For example, CPU usage or temperature.
- **Histogram**: Distribution of values with configurable buckets. For example, request latency.

Axiom doesn’t currently support the following data types:
- Exponential histograms
- `bytes`, `kvlist`, and `array` tag value types
- Exemplar, baggage, and context data
- Nanosecond-precision timestamps

### Data model optimizations

MetricsDB applies the following transformations to improve query performance and reduce storage costs:

> - **Timestamp precision**: Truncate nanosecond timestamps to second precision. Second-level granularity is sufficient for metrics analysis (unlike traces which need finer resolution), and this optimization significantly improves compression ratios and query speed.

Reviewer: ```suggestion
- **Timestamp precision**: Truncate nanosecond timestamps to second precision. MetricsDB is built for use cases where second-level granularity is sufficient, and this optimization significantly improves compression ratios and query speed.
```

import Prerequisites from "/snippets/standard-prerequisites.mdx"

Axiom’s dedicated MetricsDB provides a purpose-built metrics database that handles high-cardinality time-series data without the cost penalties and performance degradation common in traditional metrics systems. This section explains how to work with OpenTelemetry (OTel) metrics in Axiom.

<Info>
Support for OTel metrics is currently in private preview. For more information, see [Feature states](/platform-overview/roadmap#feature-states).
</Info>

## What makes MetricsDB different

MetricsDB is engineered from the ground up to embrace dimensional complexity:

> - **High-cardinality as a design principle**: Store metrics with high-cardinality dimensional tags: customer IDs, trace IDs, pod names, user agents, and other highly variable dimensions. Where other metrics databases penalize you with higher costs or degraded performance, MetricsDB treats high cardinality as a core capability.

Reviewer 2: Even if we don't suffer as much as other systems from high cardinality we should definitely not suggest storing `customer IDs` or `trace IDs` as tags, that is still an anti-pattern that will not make for a good experience either for us or the customers.

I would prefer not give any examples here as users tend to do strange things and giving an example implies we explicitly support that no matter how silly a user goes.

### Query scope

You can query one dataset per query.

### Supported data types

MetricsDB focuses on the core OpenTelemetry metric types that cover the vast majority of observability scenarios.

Axiom doesn’t currently support the following data types:
- Exponential histograms
- `bytes`, `kvlist`, and `array` tag value types
- Exemplar, baggage, and context data
- Nanosecond-precision timestamps

### Data model optimizations

MetricsDB applies the following transformations to improve query performance and reduce storage costs:

- **Timestamp precision**: Truncate nanosecond timestamps to second precision. Second-level granularity is sufficient for metrics analysis (unlike traces which need finer resolution), and this optimization significantly improves compression ratios and query speed.
- **Unified tag namespace**: Flatten resource, scope, and metric tags into a single namespace. This simplification makes queries more straightforward and enables faster dimensional filtering. You don’t need to remember which tags came from which scope.

> - **Unit normalization**: Convert the `unit` field to `otel.metric.unit` for consistent handling across all metric types.

Reviewer 2: "`unit` field" should be "`unit` attribute`"
